File: typescript/src/run_agent.ts
#!/usr/bin/env ts-node
/**
 * Production Runtime Harness — TypeScript
 * Wires react_loop.ts into a fully runnable CLI agent.
 *
 * Usage:
 *   npx ts-node run_agent.ts "your task here"
 *   npx ts-node run_agent.ts --system "You are a DevOps expert." "audit /tmp"
 *   npx ts-node run_agent.ts --max-iter 50 --json-out /tmp/result.json "task"
 *   echo "task" | npx ts-node run_agent.ts -
 *
 * Environment:
 *   ANTHROPIC_API_KEY   required
 *   AGENT_MAX_ITER      default 30
 *   AGENT_SYSTEM        system prompt override
 *   AGENT_LOG_LEVEL     debug | info (default info)
 */

import * as fs from "fs";
import * as path from "path";
import * as readline from "readline";

// ── Validate environment ──────────────────────────────────────────────────────
function validateEnv() {
  const provider = (process.env.MODEL_PROVIDER ?? "anthropic").trim().toLowerCase();
  const keyMap: Record<string, string> = {
    "anthropic":       "ANTHROPIC_API_KEY",
    "deepseek":        "DEEPSEEK_API_KEY",
    "deepseek-openai": "DEEPSEEK_API_KEY",
    "openai":          "OPENAI_API_KEY",
  };
  const envVar = keyMap[provider];
  if (!envVar) {
    process.stderr.write(`[harness] ERROR: unknown MODEL_PROVIDER '${provider}'\n`);
    process.exit(1);
  }
  const key = (process.env[envVar] ?? "").trim();
  if (!key) {
    process.stderr.write(`[harness] ERROR: ${envVar} is not set (required for provider '${provider}')\n`);
    process.exit(1);
  }
  process.stderr.write(`[harness] provider=${provider} model=${process.env.MODEL_NAME ?? "(default)"}\n`);
}

// ── Logger ────────────────────────────────────────────────────────────────────
const logLevel = (process.env.AGENT_LOG_LEVEL ?? "info").toLowerCase();
function log(level: "info" | "warn" | "error" | "debug", msg: string) {
  if (level === "debug" && logLevel !== "debug") return;
  const ts = new Date().toISOString();
  process.stderr.write(`${ts} [${level.toUpperCase()}] ${msg}\n`);
}

// ── Signal handling ───────────────────────────────────────────────────────────
let interrupted = false;
process.on("SIGINT", () => {
  if (interrupted) { process.stderr.write("\n[harness] Force quit\n"); process.exit(130); }
  interrupted = true;
  process.stderr.write("\n[harness] Interrupt — finishing current iteration...\n");
});

// ── Import engine ─────────────────────────────────────────────────────────────
import { runReact, Step } from "./react_loop";

// ── Step printer ──────────────────────────────────────────────────────────────
function makeStepPrinter(verbose: boolean) {
  if (!verbose) return undefined;
  return (step: Step) => {
    if (interrupted) throw new Error("Agent interrupted by user");
    if (step.type === "thought") {
      process.stdout.write(`\n\x1b[94m💭 THOUGHT\x1b[0m\n${step.content}\n`);
    } else if (step.type === "action") {
      const inp = JSON.stringify((step as any).input, null, 2);
      process.stdout.write(`\n\x1b[93m⚡ ACTION\x1b[0m  ${(step as any).tool}\n${inp}\n`);
    } else if (step.type === "observation") {
      const color  = (step as any).error ? "\x1b[91m" : "\x1b[92m";
      const label  = (step as any).error ? "❌ ERROR" : "👁 OBSERVE";
      const preview = step.content.slice(0, 600) + (step.content.length > 600 ? "..." : "");
      process.stdout.write(`\n${color}${label}\x1b[0m\n${preview}\n`);
    }
  };
}

// ── Arg parser ────────────────────────────────────────────────────────────────
function parseArgs() {
  const args = process.argv.slice(2);
  const opts = {
    task:     "",
    system:   process.env.AGENT_SYSTEM ?? "",
    maxIter:  parseInt(process.env.AGENT_MAX_ITER ?? "30", 10),
    jsonOut:  null as string | null,
    quiet:    false,
  };

  let i = 0;
  while (i < args.length) {
    switch (args[i]) {
      case "--system":   opts.system  = args[++i]; break;
      case "--max-iter": opts.maxIter = parseInt(args[++i], 10); break;
      case "--json-out": opts.jsonOut = args[++i]; break;
      case "--quiet":    opts.quiet   = true; break;
      default:           opts.task    = args[i]; break;
    }
    i++;
  }
  return opts;
}

// ── Read stdin ────────────────────────────────────────────────────────────────
async function readStdin(): Promise<string> {
  return new Promise((resolve) => {
    const rl = readline.createInterface({ input: process.stdin });
    const lines: string[] = [];
    rl.on("line", (l) => lines.push(l));
    rl.on("close", () => resolve(lines.join("\n").trim()));
  });
}

// ── Main ──────────────────────────────────────────────────────────────────────
async function main() {
  validateEnv();
  const opts = parseArgs();

  if (!opts.task || opts.task === "-") {
    opts.task = await readStdin();
  }
  if (!opts.task) {
    process.stderr.write("[harness] ERROR: No task provided\n");
    process.exit(1);
  }

  log("info", `Task: ${opts.task.slice(0, 200)}`);
  log("info", `Max iterations: ${opts.maxIter}`);

  const start = Date.now();

  try {
    const { answer, trace } = await runReact(opts.task, {
      systemExtra:   opts.system,
      maxIterations: opts.maxIter,
      onStep:        makeStepPrinter(!opts.quiet),
    });

    const elapsed = ((Date.now() - start) / 1000).toFixed(2);
    const sep = "=".repeat(64);

    process.stdout.write(`\n${sep}\nFINAL ANSWER\n${sep}\n${answer}\n${sep}\n`);
    process.stdout.write(`Iterations: ${trace.iterations} | Tokens: ${trace.totalTokens} | Time: ${elapsed}s\n`);

    if (opts.jsonOut) {
      const out = {
        task:        opts.task,
        answer,
        iterations:  trace.iterations,
        totalTokens: trace.totalTokens,
        elapsedS:    parseFloat(elapsed),
        steps:       trace.steps,
      };
      fs.writeFileSync(opts.jsonOut, JSON.stringify(out, null, 2), "utf8");
      log("info", `Result written to ${opts.jsonOut}`);
    }

    process.exit(0);

  } catch (e: unknown) {
    const err = e as Error;
    if (interrupted || err.message?.includes("interrupted")) {
      log("warn", "Agent stopped by user");
      process.exit(130);
    }
    log("error", `Agent failed: ${err.message}`);
    process.exit(1);
  }
}

main();

File: java/src/RunAgent.java
/**
 * Production Runtime Harness — Java
 * Wires ReactLoop.java into a fully runnable CLI agent.
 *
 * Same deps as ReactLoop.java:
 *   com.squareup.okhttp3:okhttp:4.12.0
 *   com.fasterxml.jackson.core:jackson-databind:2.17.0
 *
 * Usage:
 *   java -cp .:deps RunAgent "your task here"
 *   java -cp .:deps RunAgent --system "You are a DevOps expert." "audit /tmp"
 *   java -cp .:deps RunAgent --max-iter 50 --json-out /tmp/result.json "task"
 *   echo "task" | java -cp .:deps RunAgent -
 *
 * Environment:
 *   ANTHROPIC_API_KEY   required
 *   AGENT_MAX_ITER      default 30
 *   AGENT_SYSTEM        system prompt override
 */

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.node.ObjectNode;
import com.fasterxml.jackson.databind.node.ArrayNode;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;
import java.time.Instant;
import java.time.format.DateTimeFormatter;
import java.util.*;
import java.util.concurrent.atomic.AtomicBoolean;

public class RunAgent {

    static final ObjectMapper JSON    = new ObjectMapper();
    static final AtomicBoolean STOPPED = new AtomicBoolean(false);

    // ── Logger ────────────────────────────────────────────────────────────────
    static void log(String level, String msg) {
        String ts = DateTimeFormatter.ISO_INSTANT.format(Instant.now());
        System.err.printf("%s [%s] %s%n", ts, level, msg);
    }

    // ── Validate environment ──────────────────────────────────────────────────
    static void validateEnv() {
        String provider = System.getenv().getOrDefault("MODEL_PROVIDER", "anthropic").trim().toLowerCase();
        Map<String, String> keyMap = Map.of(
            "anthropic",       "ANTHROPIC_API_KEY",
            "deepseek",        "DEEPSEEK_API_KEY",
            "deepseek-openai", "DEEPSEEK_API_KEY",
            "openai",          "OPENAI_API_KEY"
        );
        String envVar = keyMap.get(provider);
        if (envVar == null) {
            log("ERROR", "unknown MODEL_PROVIDER '" + provider + "'");
            System.exit(1);
        }
        String key = System.getenv(envVar);
        if (key == null || key.isBlank()) {
            log("ERROR", envVar + " is not set (required for provider '" + provider + "')");
            System.exit(1);
        }
        log("INFO", "provider=" + provider + " model=" + System.getenv().getOrDefault("MODEL_NAME", "(default)"));
    }

    // ── Step printer ──────────────────────────────────────────────────────────
    static java.util.function.Consumer<ReactLoop.TraceStep> makeStepPrinter(boolean verbose) {
        if (!verbose) return step -> {};
        return step -> {
            if (STOPPED.get()) {
                throw new RuntimeException("Agent interrupted by user");
            }
            switch (step.type()) {
                case "thought" -> System.out.printf("%n\033[94m💭 THOUGHT\033[0m%n%s%n", step.content());
                case "action"  -> {
                    String inp = "";
                    try { inp = JSON.writerWithDefaultPrettyPrinter()
                            .writeValueAsString(step.input()); } catch (Exception ignored) {}
                    System.out.printf("%n\033[93m⚡ ACTION\033[0m  %s%n%s%n", step.tool(), inp);
                }
                case "observation" -> {
                    String color = step.error() ? "\033[91m" : "\033[92m";
                    String label = step.error() ? "❌ ERROR" : "👁 OBSERVE";
                    String preview = step.content().length() > 600
                        ? step.content().substring(0, 600) + "..."
                        : step.content();
                    System.out.printf("%n%s%s\033[0m%n%s%n", color, label, preview);
                }
            }
            System.out.flush();
        };
    }

    // ── Arg parser ────────────────────────────────────────────────────────────
    static class RunOptions {
        String task      = "-";
        String system    = Optional.ofNullable(System.getenv("AGENT_SYSTEM")).orElse("");
        int    maxIter   = Integer.parseInt(Optional.ofNullable(
                               System.getenv("AGENT_MAX_ITER")).orElse("30"));
        String jsonOut   = null;
        boolean quiet    = false;
    }

    static RunOptions parseArgs(String[] args) {
        RunOptions opts = new RunOptions();
        List<String> positional = new ArrayList<>();
        for (int i = 0; i < args.length; i++) {
            switch (args[i]) {
                case "--system"   -> opts.system  = args[++i];
                case "--max-iter" -> opts.maxIter = Integer.parseInt(args[++i]);
                case "--json-out" -> opts.jsonOut = args[++i];
                case "--quiet"    -> opts.quiet   = true;
                default           -> positional.add(args[i]);
            }
        }
        if (!positional.isEmpty()) {
            opts.task = String.join(" ", positional);
        }
        return opts;
    }

    // ── Read stdin ────────────────────────────────────────────────────────────
    static String readStdin() throws IOException {
        BufferedReader reader = new BufferedReader(
            new InputStreamReader(System.in, StandardCharsets.UTF_8));
        StringBuilder sb = new StringBuilder();
        String line;
        while ((line = reader.readLine()) != null) {
            sb.append(line).append("\n");
        }
        return sb.toString().trim();
    }

    // ── Signal handling ───────────────────────────────────────────────────────
    static void installShutdownHook() {
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            if (STOPPED.getAndSet(true)) {
                System.err.println("[harness] Force quit");
            } else {
                System.err.println("[harness] Interrupt — finishing current iteration...");
            }
        }));
    }

    // ── Main ──────────────────────────────────────────────────────────────────
    public static void main(String[] args) throws Exception {
        validateEnv();
        installShutdownHook();

        RunOptions opts = parseArgs(args);

        // Read task
        if ("-".equals(opts.task)) {
            opts.task = readStdin();
        }
        if (opts.task == null || opts.task.isBlank()) {
            log("ERROR", "No task provided");
            System.exit(1);
        }

        log("INFO", "Task: " + opts.task.substring(0, Math.min(200, opts.task.length())));
        log("INFO", "Max iterations: " + opts.maxIter);

        long startMs = System.currentTimeMillis();
        int exitCode = 0;

        try {
            ReactLoop.ReactResult result = ReactLoop.runReact(
                opts.task,
                opts.system.isBlank() ? null : opts.system,
                opts.maxIter,
                makeStepPrinter(!opts.quiet)
            );

            long elapsedMs = System.currentTimeMillis() - startMs;
            String sep = "=".repeat(64);

            System.out.printf("%n%s%nFINAL ANSWER%n%s%n%s%n%s%n",
                sep, sep, result.answer, sep);
            System.out.printf("Iterations: %d | Tokens: %d | Time: %.2fs%n",
                result.trace.iterations(), result.trace.totalTokens(), elapsedMs / 1000.0);

            if (opts.jsonOut != null) {
                ObjectNode doc = JSON.createObjectNode();
                doc.put("task",         opts.task);
                doc.put("answer",       result.answer);
                doc.put("iterations",   result.trace.iterations());
                doc.put("total_tokens", result.trace.totalTokens());
                doc.put("elapsed_ms",   elapsedMs);
                ArrayNode stepsArr = JSON.createArrayNode();
                result.trace.steps().forEach(s -> stepsArr.add(s.toJson()));
                doc.set("steps", stepsArr);
                Files.writeString(Path.of(opts.jsonOut),
                    JSON.writerWithDefaultPrettyPrinter().writeValueAsString(doc));
                log("INFO", "Result written to " + opts.jsonOut);
            }

        } catch (RuntimeException e) {
            if (STOPPED.get() || e.getMessage().contains("interrupted")) {
                log("WARN", "Agent stopped by user");
                exitCode = 130;
            } else {
                log("ERROR", "Agent failed: " + e.getMessage());
                exitCode = 1;
            }
        } catch (Exception e) {
            log("ERROR", "Unexpected error: " + e.getMessage());
            exitCode = 1;
        }

        System.exit(exitCode);
    }
}
